plugins/binanceExchange_ws.go

```diff
@@ -0,0 +1,283 @@
+package plugins
+
+import (
+	"fmt"
+	"log"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/adshao/go-binance/v2"
+	"github.com/stellar/kelp/api"
+	"github.com/stellar/kelp/model"
+)
+
+const (
+	STREAMTICKER = "@ticker"
+	TTLTIME      = time.Second * 3 // ttl time in seconds
+)
+
+var (
+	ErrConversionWsMarketEvent = errConversion{from: "interface", to: "*binance.WsMarketStatEvent"}
+)
+
+type errMissingSymbol struct {
+	symbol string
+}
+
+func (err errMissingSymbol) Error() string {
+	return fmt.Sprintf("Symbol %s is missing from exchange intizialization", err.symbol)
+}
+
+type errConversion struct {
+	from string
+	to   string
+}
+
+func (err errConversion) Error() string {
+	return fmt.Sprintf("Error conversion from %s to %s", err.from, err.to)
+}
+
+type stream struct {
+	doneC   chan struct{}
+	stopC   chan struct{}
+	cleanup func()
+}
+
+//Wait until the stream ends
+func (s stream) Wait() {
+
+	if s.doneC == nil {
+		return
+	}
+
+	<-s.doneC
+}
+
+//Close the stream and cleanup any data
+func (s stream) Close() {
+	if s.stopC == nil {
+		return
+	}
+	s.stopC <- struct{}{}
+	s.stopC = nil
+
+	if s.cleanup != nil {
+		s.cleanup()
+	}
+}
+
+type mapData struct {
+	data      interface{}
+	createdAt time.Time
+}
+
+func isStale(data mapData, ttl time.Duration) bool {
+
+	return time.Now().Sub(data.createdAt).Seconds() > ttl.Seconds()
+}
+
+type mapEvents struct {
+	data map[string]mapData
+	mtx  *sync.Mutex
+}
+
+func (m *mapEvents) Set(key string, data interface{}) {
+
+	now := time.Now()
+
+	m.mtx.Lock()
+	m.data[key] = mapData{
+		data:      data,
+		createdAt: now,
+	}
+	m.mtx.Unlock()
```

Comment:
generally prefer using the defer pattern like this:
```golang
lock.Lock()
defer lock.Unlock()
```

this is a pattern we use across the codebase and is generally safer (without going into details here).

Reply:
https://github.com/golang/go/issues/14939, I can do that, but we have to take in consideration the performance issue from what I know this was improved in go 1.14+
![image](https://user-images.githubusercontent.com/37719049/126064869-803fd479-01a1-48d6-86a5-bbb7ccdfac31.png)
https://golang.org/doc/go1.14, but the project is using go 1.13

Reply:
discussed in our chat. `defer` for now since we will be upgrading beyond go 1.13 soon